logstash-core/src/main/java/org/logstash/execution/AbstractPipelineExt.java

```diff
@@ -593,6 +583,71 @@ private UptimeMetric initOrGetUptimeMetric(final ThreadContext context,
+        if (!queueContext.isNil() && queueContext.asJavaString().equals(QueueFactoryExt.PERSISTED_TYPE)) {
+
+            final RubySymbol[] queueNamespace = buildNamespace(QUEUE_KEY);
+            final RubySymbol[] queueCapacityNamespace = buildNamespace(QUEUE_KEY, CAPACITY_KEY);
+
+            final Supplier<NumberGauge> eventsGaugeMetricSupplier = () -> initOrGetNumberGaugeMetric(context, queueNamespace, EVENTS_KEY).orElse(null);
+            final FlowMetric growthEventsFlow = createFlowMetric(QUEUE_PERSISTED_GROWTH_EVENTS_KEY, eventsGaugeMetricSupplier, () -> uptimeInPreciseSeconds);
+            this.flowMetrics.add(growthEventsFlow);
+            storeMetric(context, flowNamespace, growthEventsFlow);
+
+            final Supplier<NumberGauge> queueSizeInBytesMetricSupplier = () -> initOrGetNumberGaugeMetric(context, queueCapacityNamespace, QUEUE_SIZE_IN_BYTES_KEY).orElse(null);
+            final FlowMetric growthBytesFlow = createFlowMetric(QUEUE_PERSISTED_GROWTH_BYTES_KEY, queueSizeInBytesMetricSupplier, () -> uptimeInPreciseSeconds);
+            this.flowMetrics.add(growthBytesFlow);
+            storeMetric(context, flowNamespace, growthBytesFlow);
+        }
+    }
+
+    private void initializePluginFlowMetrics(final ThreadContext context, final Metric<Number> uptimeInPreciseMillis) {
+        for (IRubyObject rubyObject: lirExecution.inputs()) {
+            IRubyObject rubyfiedId = rubyObject.callMethod(context, "id");
+            String id = rubyfiedId.toString();
+            initializePluginThroughputFlowMetric(context, uptimeInPreciseMillis, id);
+        }
+
+        final int workerCount = getSetting(context, SettingKeyDefinitions.PIPELINE_WORKERS).convertToInteger().getIntValue();
+
+        for (AbstractFilterDelegatorExt delegator: lirExecution.filters()) {
+            initializePluginWorkerFlowMetrics(context, workerCount, uptimeInPreciseMillis, FILTERS_KEY, delegator.getId().asJavaString());
+        }
+
+        for (AbstractOutputDelegatorExt delegator: lirExecution.outputs()) {
+            initializePluginWorkerFlowMetrics(context, workerCount, uptimeInPreciseMillis, OUTPUTS_KEY, delegator.getId().asJavaString());
+        }
+    }
+
+    private void initializePluginThroughputFlowMetric(final ThreadContext context, final Metric<Number> uptimeInPreciseMillis, final String id) {
+        final RubySymbol[] eventsNamespace = buildNamespace(PLUGINS_KEY, INPUTS_KEY, RubyUtil.RUBY.newSymbol(id), EVENTS_KEY);
+        final LongCounter counter = initOrGetCounterMetric(context, eventsNamespace, OUT_KEY);
+
+        final FlowMetric throughputFlow = createFlowMetric(PLUGIN_THROUGHPUT_KEY, counter, uptimeInPreciseMillis);
+        this.flowMetrics.add(throughputFlow);
+
+        final RubySymbol[] flowNamespace = buildNamespace(PLUGINS_KEY, INPUTS_KEY, RubyUtil.RUBY.newSymbol(id), FLOW_KEY);
+        storeMetric(context, flowNamespace, throughputFlow);
+    }
```

Comment:
throughput is expressed in events-per-second, not events-per-milli.

```suggestion
    private void initializePluginThroughputFlowMetric(final ThreadContext context, final UptimeMetric uptime) {
        final Metric<Number> uptimeInPreciseSeconds = uptimeMetric.withUnitsPrecise(SECONDS);
        final RubySymbol[] eventsNamespace = buildNamespace(PLUGINS_KEY, INPUTS_KEY, RubyUtil.RUBY.newSymbol(id), EVENTS_KEY);
        final LongCounter counter = initOrGetCounterMetric(context, eventsNamespace, OUT_KEY);

        final FlowMetric throughputFlow = createFlowMetric(PLUGIN_THROUGHPUT_KEY, counter, uptimeInPreciseSeconds);
        this.flowMetrics.add(throughputFlow);

        final RubySymbol[] flowNamespace = buildNamespace(PLUGINS_KEY, INPUTS_KEY, RubyUtil.RUBY.newSymbol(id), FLOW_KEY);
        storeMetric(context, flowNamespace, throughputFlow);
    }
```